File: dotnet/src/Microsoft.Agents.Protocol.Validation/ValidationResult.cs
namespace Microsoft.Agents.Protocol.Validation;

/// <summary>
/// Represents the result of a validation operation.
/// </summary>
public class ValidationResult
{
    /// <summary>
    /// Gets a value indicating whether the validation passed (no errors).
    /// </summary>
    public bool IsValid => !Errors.Any(e => e.Severity == ValidationSeverity.Error);

    /// <summary>
    /// Gets the list of validation errors.
    /// </summary>
    public IReadOnlyList<ValidationError> Errors { get; }

    /// <summary>
    /// Gets a value indicating whether there are any warnings.
    /// </summary>
    public bool HasWarnings => Errors.Any(e => e.Severity == ValidationSeverity.Warning);

    /// <summary>
    /// Creates a new successful validation result with no errors.
    /// </summary>
    public ValidationResult()
    {
        Errors = Array.Empty<ValidationError>();
    }

    /// <summary>
    /// Creates a new validation result with the specified errors.
    /// </summary>
    public ValidationResult(IEnumerable<ValidationError> errors)
    {
        Errors = errors.ToList().AsReadOnly();
    }

    /// <summary>
    /// Creates a new validation result with a single error.
    /// </summary>
    public ValidationResult(ValidationError error)
    {
        Errors = new[] { error }.ToList().AsReadOnly();
    }

    /// <summary>
    /// Combines this validation result with another validation result.
    /// </summary>
    /// <param name="other">The other validation result to combine with.</param>
    /// <returns>A new validation result containing errors from both results.</returns>
    public ValidationResult Combine(ValidationResult other)
    {
        if (other == null)
            return this;

        var combinedErrors = Errors.Concat(other.Errors).ToList();
        return new ValidationResult(combinedErrors);
    }

    /// <summary>
    /// Throws a ValidationException if this result is not valid.
    /// </summary>
    public void ThrowIfInvalid()
    {
        if (!IsValid)
        {
            throw new ValidationException(this);
        }
    }

    /// <summary>
    /// Gets all errors of the specified severity.
    /// </summary>
    public IEnumerable<ValidationError> GetErrorsBySeverity(ValidationSeverity severity)
    {
        return Errors.Where(e => e.Severity == severity);
    }

    /// <summary>
    /// Gets all errors with the specified code.
    /// </summary>
    public IEnumerable<ValidationError> GetErrorsByCode(string code)
    {
        return Errors.Where(e => e.Code == code);
    }

    /// <summary>
    /// Creates a successful validation result.
    /// </summary>
    public static ValidationResult Success()
    {
        return new ValidationResult();
    }

    /// <summary>
    /// Creates a validation result with a single error.
    /// </summary>
    public static ValidationResult Failure(string code, string message, string? field = null)
    {
        return new ValidationResult(new ValidationError(code, message, field));
    }

    /// <summary>
    /// Returns a string representation of this validation result.
    /// </summary>
    public override string ToString()
    {
        if (IsValid)
            return "Validation passed";

        var errorCount = GetErrorsBySeverity(ValidationSeverity.Error).Count();
        var warningCount = GetErrorsBySeverity(ValidationSeverity.Warning).Count();

        return $"Validation failed: {errorCount} error(s), {warningCount} warning(s)";
    }
}

File: typescript/packages/agents-protocol-validation/src/core/ValidationResult.ts
import { ValidationError, ValidationSeverity } from './ValidationError';

/**
 * Represents the result of a validation operation.
 */
export class ValidationResult {
  /**
   * Gets the list of validation errors.
   */
  public readonly errors: ReadonlyArray<ValidationError>;

  constructor(errors: ValidationError[] = []) {
    this.errors = errors;
  }

  /**
   * Gets a value indicating whether the validation passed (no errors).
   */
  get isValid(): boolean {
    return !this.errors.some(e => e.severity === ValidationSeverity.Error);
  }

  /**
   * Gets a value indicating whether there are any warnings.
   */
  get hasWarnings(): boolean {
    return this.errors.some(e => e.severity === ValidationSeverity.Warning);
  }

  /**
   * Combines this validation result with another validation result.
   */
  combine(other: ValidationResult): ValidationResult {
    if (!other) {
      return this;
    }
    return new ValidationResult([...this.errors, ...other.errors]);
  }

  /**
   * Throws a ValidationException if this result is not valid.
   */
  throwIfInvalid(): void {
    if (!this.isValid) {
      throw new ValidationException(this);
    }
  }

  /**
   * Gets all errors of the specified severity.
   */
  getErrorsBySeverity(severity: ValidationSeverity): ValidationError[] {
    return this.errors.filter(e => e.severity === severity);
  }

  /**
   * Gets all errors with the specified code.
   */
  getErrorsByCode(code: string): ValidationError[] {
    return this.errors.filter(e => e.code === code);
  }

  /**
   * Creates a successful validation result.
   */
  static success(): ValidationResult {
    return new ValidationResult();
  }

  /**
   * Creates a validation result with a single error.
   */
  static failure(code: string, message: string, field?: string): ValidationResult {
    return new ValidationResult([new ValidationError(code, message, field)]);
  }

  /**
   * Returns a string representation of this validation result.
   */
  toString(): string {
    if (this.isValid) {
      return 'Validation passed';
    }

    const errorCount = this.getErrorsBySeverity(ValidationSeverity.Error).length;
    const warningCount = this.getErrorsBySeverity(ValidationSeverity.Warning).length;

    return `Validation failed: ${errorCount} error(s), ${warningCount} warning(s)`;
  }
}

/**
 * Exception thrown when validation fails.
 */
export class ValidationException extends Error {
  public readonly validationResult: ValidationResult;

  constructor(validationResult: ValidationResult) {
    super(ValidationException.formatMessage(validationResult));
    this.name = 'ValidationException';
    this.validationResult = validationResult;
    Object.setPrototypeOf(this, ValidationException.prototype);
  }

  private static formatMessage(result: ValidationResult): string {
    const errors = result.getErrorsBySeverity(ValidationSeverity.Error);
    if (errors.length === 0) {
      return 'Validation failed';
    }

    if (errors.length === 1) {
      return `Validation failed: ${errors[0]}`;
    }

    const errorList = errors.map((e, i) => `  ${i + 1}. ${e}`).join('\n');
    return `Validation failed with ${errors.length} error(s):\n${errorList}`;
  }
}
